### PR TITLE
[Concurrency] Add "async" operation for continuing work asynchronously.

### PR DESCRIPTION
**Pull request by DougGregor**
The `async` operation is a global function that initiates asynchronous
work on behalf of the synchronous code that calls it. Unlike `detach`,
`async` inherits priority, actor context, and other aspects of the
synchronous code that initiates it, making it a better "default"
operation for creating asynchronous work than `detach`. The `detach`
operation is still important for creating truly detached tasks that
can later be `await`'d or cancelled if needed.

Implements the main entry point for rdar://76927008.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/Actor.cpp [hunk 1]
```diff
@@ -259,6 +260,13 @@ static bool isExecutingOnMainThread() {
 #endif
 }
 
+JobPriority swift::swift_task_getCurrentThreadPriority() {
+  if (isExecutingOnMainThread())
+    return JobPriority::UserInitiated;
+
+  return static_cast<JobPriority>(qos_class_self());
+}
```
Comment (ktoso) [inline stdlib/public/Concurrency/Actor.cpp:268]: oh, nice :)

File: stdlib/public/Concurrency/Task.swift [hunk 2]
```diff
@@ -481,6 +481,55 @@ public func detach<T>(
...
+/// - Parameters:
+///   - priority: priority of the task. If unspecified, the priority will
+///               be inherited from the task that is currently executing
+///               or, if there is none, from the platform's understanding of
+///               which thread is executing.
+///   - operation: the operation to execute
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public func async(
+  priority: Task.Priority = .unspecified,
+  @_inheritActorContext @_implicitSelfCapture operation: __owned @Sendable @escaping () async -> Void
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:503]: oh both that's a lot of annotations, but love how precisely this expresses stuff :)

File: stdlib/public/Concurrency/Task.swift [hunk 3]
```diff
@@ -481,6 +481,55 @@ public func detach<T>(
...
+  let actualPriority: Task.Priority
+  if priority == .unspecified {
+    actualPriority = withUnsafeCurrentTask { task in
+      // If we are running on behalf of a task,
+      if let task = task {
+        return task.priority
+      }
+
+      return Task.Priority(rawValue: _getCurrentThreadPriority()) ?? .unspecified
+    }
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:515]: 👍

**Review by ktoso — APPROVED** [review]
Ah cool, not much magic at all then. Very nice 👍 

I'll implement propagating locals through these once it merges 👍

File: stdlib/public/Concurrency/Task.swift [hunk 4]
```diff
@@ -481,6 +481,55 @@ public func detach<T>(
...
+/// does not return a handle to refer to the task.
+///
+/// - Parameters:
+///   - priority: priority of the task. If unspecified, the priority will
+///               be inherited from the task that is currently executing
+///               or, if there is none, from the platform's understanding of
+///               which thread is executing.
+///   - operation: the operation to execute
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public func async(
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:501]: Man... I'd really like a verb here; Isn't this a `send`?

- spawn - make child task, will await it
- detach - completely separate; could wait if one wanted to, carries no values/things implicitly
- send - fire and forget, cannot await on the result; usable from sync code, does carry priority and locals <- this is exactly what doing a "fire and forget" message send would want to be, so I'm so tempted to call it a send... `async` is a bit weird and it's dispatch wording inheritance too..

Comment (hassila) [reply stdlib/public/Concurrency/Task.swift:501]: > Man... I'd really like a verb here; Isn't this a `send`?

Nah, it's a `dispatch` ;-)
(yeah, gcd connotations I know, but one could say the same for `send(2)`)

File: stdlib/public/Concurrency/Task.swift [hunk 5]
```diff
@@ -481,6 +481,55 @@ public func detach<T>(
   return Task.Handle<T, Error>(task)
 }
 
+/// Run given `operation` as asynchronously in its own top-level task.
+///
+/// The `async` function should be used when creating asynchronous work
+/// that operations on behalf of the synchronous function that calls it.
```
Comment (JensAyton) [inline stdlib/public/Concurrency/Task.swift:487]: nit: “operates”

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**